File: probs/usaco/cowcode/cowcode.cpp
#include <bits/stdc++.h>

using namespace std;

string s;
long long n, m;
int main()
{
	freopen("cowcode.in", "r", stdin);
	freopen("cowcode.out", "w", stdout);
	cin >> s >> n;
	m = s.length();
	while(m * 2 < n) m *= 2;
	while(n > s.length())
	{
		// ------------------------------------------------ //
		// remap n to first half but remember cyclic shift. //
		// m+1 goes to m, and the rest go to 1...m-1.       //
		// ------------------------------------------------ //
		if(n == m+1) n = m;
		else n -= m+1;

		// decrease m to be suitable for the next step
		// sums to lg n operations over all indices
		while(m >= n) m /= 2;
	}
	// O(lg n) time in total
	cout << s[n-1] << endl; // guaranteed to be under n now
}

File: probs/usaco/cowcode/cowcode.py
"""
Solution by Albert Ye
AlphaStar 2020 Summer Camp: CC31AB
Problem: jan17-cowcode
"""
s, n = input().split(' ')
n = int(n)
m = len(s)
while m*2 < n :
    m *= 2
while n > len(s) :
    ###################################################
    # remap n to first half but remember cyclic shift #
    # m+1 maps to m and m+2...2m go to 1...m-1.       #
    ###################################################
    if n == m+1 :
        n = m
    else :
        n -= m+1
    ###################################################
    # decrease m for the next step                    #
    ###################################################
    while m >= n :
        m //= 2

# total complexity: O(lg n)
print(s[n-1])
